### PR TITLE
Cluster restart model auto redeploy

### PR DESCRIPTION
**Pull request by zane-neo**
### Description
Support cluster level restart model auto redeploy.
 
### Issues Resolved
https://github.com/opensearch-project/ml-commons/issues/1641
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java [hunk 1]
```diff
@@ -685,6 +686,7 @@ public List<Setting<?>> getSettings() {
                 MLCommonsSettings.ML_COMMONS_ENABLE_INHOUSE_PYTHON_MODEL,
                 MLCommonsSettings.ML_COMMONS_MODEL_AUTO_REDEPLOY_ENABLE,
                 MLCommonsSettings.ML_COMMONS_MODEL_AUTO_REDEPLOY_LIFETIME_RETRY_TIMES,
+                MLCommonsSettings.ML_COMMONS_MODEL_AUTO_REDEPLOY_SUCCESS_RATIO,
```
Comment (Zhangxunmt) [inline plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java:689]: This setting is not used anywhere in this PR?

**Review by Zhangxunmt — COMMENTED** [review]
Is the flag plugins.ml_commons.model_auto_redeploy.enable for cluster level or node level after this change?

File: plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java (hunk 1 above)
Comment (zane-neo) [reply plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java:689]: Didn't expose this setting when implementing the auto redeploy feature, so exposing it now to enable customer to change this ratio value.

File: plugin/src/main/java/org/opensearch/ml/cluster/MLCommonsClusterManagerEventListener.java [hunk 2]
```diff
@@ -62,13 +70,28 @@ public MLCommonsClusterManagerEventListener(
...
+            if (syncModelRoutingCron == null) {
+                startSyncModelRoutingCron();
+            }
+        });
+        mlModelAutoReDeployer.setStartCronJobListener(listener);
+        String localNodeId = clusterService.localNode().getId();
+        threadPool
+            .schedule(
+                () -> mlModelAutoReDeployer.buildAutoReloadArrangement(List.of(localNodeId), localNodeId),
+                TimeValue.timeValueSeconds(jobInterval),
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/cluster/MLCommonsClusterManagerEventListener.java:87]: Can we start auto reload immediately ?

Comment (zane-neo) [reply plugin/src/main/java/org/opensearch/ml/cluster/MLCommonsClusterManagerEventListener.java:87]: No, cluster manager ready doesn't mean cluster is ready, if we start auto reload immediately we can get a `cluster not initialized` etc exception when querying needs reload models.

**Comment by zane-neo** [comment]
> Is the flag plugins.ml_commons.model_auto_redeploy.enable for cluster level or node level after this change?

Both level, this change is more like an enhancement to model auto redeploy to support cluster level.

**Review by ylwu-amzn — APPROVED** [review]

**Comment by dhrubo-os** [comment]
Just curious to know if this will also re-deploy the hidden model (what we are building now? Currently only domain manager with having the admin SSL certificate can deploy the hidden model.

**Review by Zhangxunmt — APPROVED** [review]

**Comment by ylwu-amzn** [comment]
> Just curious to know if this will also re-deploy the hidden model (what we are building now? Currently only domain manager with having the admin SSL certificate can deploy the hidden model.

I think we have no way to support hidden model auto redeployment, domain manager should build some logic outside of cluster.

**Pull request merged**